currency_monthly_rate/models/res_currency.py

```diff
@@ -0,0 +1,95 @@
+# Copyright 2018 Camptocamp SA
+# License AGPL-3.0 or later (http://www.gnu.org/licenses/agpl).
+
+from odoo import models, fields, api
+
+
+class Currency(models.Model):
+    _inherit = "res.currency"
+
+    monthly_rate = fields.Float(compute='_compute_current_monthly_rate',
+                                string='Current Monthly Rate', digits=(12, 6),
+                                help='The monthly rate of the currency to '
+                                     'the currency of rate 1.')
+    monthly_rate_ids = fields.One2many('res.currency.rate.monthly',
+                                       'currency_id', string='Monthly rates')
+
+    @api.multi
+    def _compute_current_monthly_rate(self):
+        date = self._context.get('date') or fields.Date.today()
+        company_id = self._context.get('company_id') or self.env[
+            'res.users']._get_company().id
+        # the subquery selects the first rate before 'date' for the given
+        # currency/company
+        query = """SELECT c.id, (SELECT r.rate
+                                 FROM res_currency_rate_monthly r
+                                 WHERE r.currency_id = c.id
+                                 AND r.name <= %s
+                                 AND (r.company_id IS NULL
+                                      OR r.company_id = %s)
+                                 ORDER BY r.company_id, r.name DESC
+                                 LIMIT 1) AS rate
+                   FROM res_currency c
+                   WHERE c.id IN %s"""
+        self._cr.execute(query, (date, company_id, tuple(self.ids)))
+        currency_rates = dict(self._cr.fetchall())
+        for currency in self:
+            currency.monthly_rate = currency_rates.get(currency.id) or 1.0
+
+    @api.model
+    def _get_conversion_rate(self, from_currency, to_currency):
+        monthly = self._context.get('monthly_rate')
+        if not monthly:
+            return super(Currency, self)._get_conversion_rate(from_currency,
+                                                              to_currency)
+        from_currency = from_currency.with_env(self.env)
+        to_currency = to_currency.with_env(self.env)
+        return to_currency.monthly_rate / from_currency.monthly_rate
+
+
+class CurrencyRateMonthly(models.Model):
+
+    _inherit = "res.currency.rate"
+    _name = "res.currency.rate.monthly"
+    _description = "Currency monthly rate"
+
+    def _default_get_month(self):
+        return fields.Date.from_string(
+            fields.Date.context_today(self)).strftime('%m')
+
+    def _default_get_year(self):
+        return fields.Date.from_string(
+            fields.Date.context_today(self)).strftime('%Y')
+
+    name = fields.Date(compute='_compute_name', store=True, required=True,
+                       index=True)
+    year = fields.Char(size=4, required=True, default=_default_get_year)
```

Comment:
proxy the method w/ a lambda

Reply:
I'm not sure it's really worth it as we loose a bit in readability :
```
    year = fields.Char(size=4, required=True,
                       default=lambda self: fields.Date.from_string(
                           fields.Date.context_today(self)).strftime('%m'))
```
vs 
```
    def _default_get_year(self):
        return fields.Date.from_string(
            fields.Date.context_today(self)).strftime('%Y')

    year = fields.Char(size=4, required=True, default=_default_get_year)
```
:confused:

Reply:
is not about readability, is about to make it easy to override. If you reference the method you are referencing that precise method object so that if you want to override the default you have to override the whole field. And w/ "proxy w/ lambda" I meant:
```python
default=lambda self: self._default_get_year()
```
which is readable as the other one :wink:

Reply:
Makes sense, thanks for the explanation.